pkg/plugins/policies/meshretry/plugin/xds/configurer.go

```diff
@@ -0,0 +1,272 @@
+package xds
+
+import (
+	"net/http"
+	"strconv"
+	"strings"
+
+	envoy_listener "github.com/envoyproxy/go-control-plane/envoy/config/listener/v3"
+	envoy_route "github.com/envoyproxy/go-control-plane/envoy/config/route/v3"
+	envoy_hcm "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/network/http_connection_manager/v3"
+	envoy_tcp "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/network/tcp_proxy/v3"
+	envoy_type_matcher "github.com/envoyproxy/go-control-plane/envoy/type/matcher/v3"
+
+	"github.com/kumahq/kuma/api/common/v1alpha1"
+	core_mesh "github.com/kumahq/kuma/pkg/core/resources/apis/mesh"
+	api "github.com/kumahq/kuma/pkg/plugins/policies/meshretry/api/v1alpha1"
+	util_proto "github.com/kumahq/kuma/pkg/util/proto"
+	v3 "github.com/kumahq/kuma/pkg/xds/envoy/listeners/v3"
+)
+
+const (
+	HttpRetryOnDefault = "gateway-error,connect-failure," +
+		"refused-stream"
+	HttpRetryOnRetriableStatusCodes = "retriable-status-codes"
+	GrpcRetryOnDefault              = "cancelled,connect-failure," +
+		"gateway-error,refused-stream,reset,resource-exhausted,unavailable"
+)
+
+type Configurer struct {
+	Retry    *api.Conf
+	Protocol core_mesh.Protocol
+}
+
+func genGrpcRetryPolicy(conf *api.GRPC) *envoy_route.RetryPolicy {
+	if conf == nil {
+		return nil
+	}
+
+	policy := envoy_route.RetryPolicy{
+		RetryOn: GrpcRetryOnDefault,
+	}
+
+	if conf.PerTryTimeout != nil {
+		policy.PerTryTimeout = util_proto.Duration(conf.PerTryTimeout.Duration)
+	}
+
+	if conf.NumRetries != nil {
+		policy.NumRetries = util_proto.UInt32(*conf.NumRetries)
+	}
+
+	retryOn := GrpcRetryOn(conf.RetryOn)
+	if retryOn != "" {
+		policy.RetryOn = retryOn
+	}
+
+	if conf.BackOff != nil {
+		policy.RetryBackOff = &envoy_route.RetryPolicy_RetryBackOff{}
+
+		if conf.BackOff.BaseInterval != nil {
+			policy.RetryBackOff.BaseInterval = util_proto.Duration(conf.BackOff.BaseInterval.Duration)
+		}
+		if conf.BackOff.MaxInterval != nil {
+			policy.RetryBackOff.MaxInterval = util_proto.Duration(conf.BackOff.MaxInterval.Duration)
+		}
+	}
+
+	return &policy
+}
+
+func genHttpRetryPolicy(conf *api.HTTP) *envoy_route.RetryPolicy {
+	if conf == nil {
+		return nil
+	}
+
+	policy := envoy_route.RetryPolicy{
+		RetryOn: HttpRetryOnDefault,
+	}
+
+	if conf.PerTryTimeout != nil {
+		policy.PerTryTimeout = util_proto.Duration(conf.PerTryTimeout.Duration)
+	}
+
+	if conf.NumRetries != nil {
+		policy.NumRetries = util_proto.UInt32(*conf.NumRetries)
+	}
+
+	if conf.BackOff != nil {
```

Comment:
we are missing one imporatnt in my opinion functionality. We cannot configure to take into account `Retry-After` header. Example [here](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#envoy-v3-api-msg-config-route-v3-retrypolicy-ratelimitedretrybackoff). It is expecially important when you plan on setting default retry policy for Mesh. In Allegro we have lot's of services that were respecting this header and after applying default retry we broke this behaviour. I assume that this is the case for bigger and more mature organisations, but it would be nice to include this

Reply:
@lobkovilya, @lahabana - WDYT? should I implement this or is this something to create an issue about implement when needed?

Reply:
Added, we can always remove it later.